[CS-4682] Connecting new backup flow into onboarding

oidoug opened the pull request:
### Description

This PR adds the backup flow in onboarding by updating the logic in `useShowOnboarding` hook. It changes some auxiliary code but the essence is in the hook.

It also adds a new persisted preference flag in case uses skip the backup flow.

**Note:** I added a AsyncStore.clear() on resetting the wallet so preferences stored with redux-persist, like skip backup or profile, would also be cleared. Let me know if you remember something that needs to be maintained over resets.

- [x] Completes #(CS-4682)

### Checklist

- [x] Tested on a small device
- [x] Tested on iOS
- [x] Tested on Android

### Screenshots

Several recording are available in Linear [here.](https://linear.app/cardstack/issue/CS-4682/[logic]-connect-new-backup-flow-to-the-onboarding-experience)

oidoug committed:
Adds backup into onboarding flow;
Updates backup navigation parameters to include pop stack property;
Updates backup screens to handle both being in onboard or opened from settings;
Adds persisted flag to skip backup;
Fixes reset wallet function to also clear AsyncStorage data (redux-persist preferences);

oidoug committed:
Improves test on phrase confirmation;

oidoug committed:
Adds more and fixes tests to useShowOnboarding;

paulinhapenedo commented on src/model/wallet.ts:
```diff
@@ -904,6 +905,9 @@ export const resetWallet = async () => {
       await wipeSecureStorage(allWallets);
       await keychain.wipeKeychain();
 
+      // clear preferences on redux persist properties
+      await AsyncStorage.clear();
+
```
Not sure about this approach, because we don't seem to be interacting with Async Storage directly anymore, only through Redux.

That's why we have `store.dispatch(skipProfileCreation(false));` a few lines below this code and why the actions receive params (true/false).

Maybe we could create an action that resets both `persistedFlags` and call it here?

oidoug replied:
I totally missed that!

Good idea, I'll add a "clearFlags" action then.

oidoug replied:
These dispatching for skipProfileCreation and skipBackup actions are not saving the values in local store in time before the wallet reset happens. Not sure why yet, but investigating.

paulinhapenedo commented on cardstack/src/screens/Backup/BackupExplanationScreen/BackupExplanationScreen.tsx:
```diff
@@ -14,26 +10,31 @@ import {
 
   const handleLaterOnPress = useCallback(() => {
-    navDispatch(StackActions.popToTop());
-  }, [navDispatch]);
+    triggerSkipBackup();
+  }, [triggerSkipBackup]);
```
Right now, tapping on it does nothing on the screen. We could add either the `StackActions.popToTop()` like before (which flashes the Wallet screen before moving to the Profile Creation flow) or `navigate(Routes.PROFILE_SLUG)` directly.

oidoug replied:
Strange, this should be navigating to profile through the effect on `onShowOnboarding`, it watches `selectedWallet.manuallyBackedUp` and `hasSkippedBackup`, so if any changes it should navigate to profile.

Something else must be off, I'll ping you to discuss offline.

paulinhapenedo commented on cardstack/src/screens/Backup/BackupSeedPhraseConfirmationScreen/__tests__/useBackupSeedPhraseConfirmationScreen.test.ts:
```diff
@@ -30,10 +29,11 @@ const mockSeedPhrase =
 
 describe('BackupSeedPhraseConfirmationScreen', () => {
-  const mockParams = () => {
+  const mockParams = (overwrite: any = {}) => {
```
Couldn't this be `BackupRouteParams` instead of `any`?

oidoug replied:
yes well observed, I'll update it.

paulinhapenedo commented on cardstack/src/screens/Backup/BackupSeedPhraseConfirmationScreen/useBackupSeedPhraseConfirmationScreen.ts:
```diff
@@ -54,8 +56,11 @@ export const useBackupSeedPhraseConfirmationScreen = () => {
 
   const handleConfirmPressed = useCallback(() => {
     confirmBackup();
-    navigate(Routes.WALLET_SCREEN);
-  }, [navigate, confirmBackup]);
+
+    if (popStackOnSuccess) {
+      navDispatch(StackActions.pop(popStackOnSuccess));
+    }
+  }, [confirmBackup, popStackOnSuccess, navDispatch]);
```
Tapping the "Done" button does nothing on the screen. We have to add the navigation here again. 

Also, keep in mind that we are reusing this function on line 70 of this file, in the `handleBackupToCloudPress` function. The user only sees the "success" buttons (Done and Backup to Cloud) after successfully choosing the words. If they tap on the latter, we have to set the manual backup as `true` and also redirect them to the cloud password screen.

paulinhapenedo commented on cardstack/src/hooks/onboarding/__tests__/useShowOnboarding.test.ts:
```diff
@@ -29,51 +35,47 @@ jest.mock('@cardstack/redux/persistedFlagsSlice', () => ({
-  }));
-
 describe('useShowOnboarding', () => {
+  const mockUseWallets = (overwriteParams: any = {}) => {
```
nit: we could replace the `any` with `Partial<ReturnType<typeof useWallets>>`

paulinhapenedo commented on cardstack/src/hooks/onboarding/__tests__/useShowOnboarding.test.ts:
```diff
@@ -29,51 +35,47 @@ jest.mock('@cardstack/redux/persistedFlagsSlice', () => ({
+    }));
+  };
+
+  const mockPrimarySafeHelper = (overwriteParams: any = {}) =>
```
nit: we could replace the `any` with `Partial<ReturnType<typeof usePrimarySafe>>`

danibonilha commented on cardstack/src/hooks/onboarding/__tests__/useShowOnboarding.test.ts:
```diff
@@ -29,51 +35,47 @@ jest.mock('@cardstack/redux/persistedFlagsSlice', () => ({
-
-  it('should NOT navigate to profile flow when profile hasnt been fetched', () => {
-    mockPrimarySafeHelper({ hasProfile: false, isLoadingOnInit: true });
-
```
I'm confused on why these tests were switched, and we removed this second render test, it is important to test  rerendering on this case, because this hook leaves on the wallet root

oidoug replied:
The updated hook implementation has a similar test using the callback.

danibonilha commented on cardstack/src/hooks/onboarding/useShowOnboarding.ts:
```diff
@@ -6,15 +6,32 @@ import { useAuthSelector } from '@cardstack/redux/authSlice';
   const { hasWallet } = useAuthSelector();
 
   useEffect(() => {
+    // Wait until wallet and backup info has properly being loaded.
```
nit: wait until wallet and backup info is loaded

oidoug committed:
Fixes navigation issues in onboarding;
Fixes clearing preference flags from persistance;
Adds navigateToNextOnboardingStep callback to useShowOnboading hook to control navigation with an option to navigate specifically to a screen.

oidoug committed:
Removes debug log

oidoug committed:
Updates tests for useShowOnboarding and useBackupSeedPhraseConfirmationScreen.

oidoug committed:
Re-adds test for when safes are still loading on show onboarding hook.

oidoug commented on cardstack/src/redux/persistedFlagsSlice.ts:
```diff
@@ -19,12 +21,23 @@ const slice = createSlice({
+    clearFlags(state) {
+      state = initialState;
+
+      // Hack note: Returning the state in slice reducers is not necessary generally,
```
```suggestion
      // Hack note: Returning the state in slice reducers is generally not necessary,
```

oidoug committed:
Updates BackupExplanationScreen to use new useSelectedWallet

paulinhapenedo commented on cardstack/src/screens/Backup/BackupExplanationScreen/BackupExplanationScreen.tsx:
```diff
@@ -13,27 +9,35 @@ import {
-    navDispatch(StackActions.popToTop());
-  }, [navDispatch]);
+    triggerSkipBackup();
+    navigateToNextOnboardingStep(Routes.PROFILE_SLUG);
```
If I'm importing an account that already has a profile, we don't want to be sending the user to the profile creation flow. Maybe we should not pass a route here?

oidoug replied:
Hum.. good point, problem is that this route here has a purpose. `triggerSkipBackup` will not update the `skipBackup` flag in time when `navigateToNextOnboardingStep` is called, so it will not navigate even when it needs to. 🤔

oidoug committed:
Sets manuallyBackedUp on imported wallets.

oidoug committed:
Fix for not showing profile flow on import when user already has one.

oidoug committed:
Fix: when importing, backup flow will not show so WalletScreen needs to wait for safes to load before trying to show profile flow.

paulinhapenedo commented on cardstack/src/redux/persistedFlagsSlice.ts:
```diff
@@ -19,12 +21,23 @@ const slice = createSlice({
+      state.hasSkippedBackup = action.payload;
+    },
+    clearFlags(state) {
+      state = initialState;
+
+      // Hack note: Returning the state in slice reducers is not necessary generally,
+      // but this state does not update when the action is dispatched
+      // from resetWallet in wallet model for some unknown reason.
+      return state;
```
You can only return the `initialState` as mentioned in the [docs](https://redux-toolkit.js.org/usage/immer-reducers#resetting-and-replacing-state). There's no need to assign it to `state`.

oidoug replied:
Thanks for the pointer! I was wondering why in some places we had to return and others not.

paulinhapenedo reviewed (approved):
I believe we can merge and handle the navigation issues in another PR.

oidoug committed:
Removed redudant set state in persistedFlagsSlice.
